modules/placement-driver/src/main/java/org/apache/ignite/internal/placementdriver/leases/LeaseTracker.java

```diff
@@ -152,4 +195,46 @@ public CompletableFuture<Void> onUpdate(WatchEvent event) {
         public void onError(Throwable e) {
         }
     }
+
+    @Override
+    public CompletableFuture<LeaseMeta> awaitPrimaryReplica(ReplicationGroupId groupId, HybridTimestamp timestamp) {
+        if (!busyLock.enterBusy()) {
+            return failedFuture(new NodeStoppingException("Component is stopping."));
+        }
+        try {
+            return primaryReplicaWaiters.computeIfAbsent(groupId, id -> new PendingIndependentComparableValuesTracker<>(MIN_VALUE))
+                    .waitFor(timestamp);
+        } finally {
+            busyLock.leaveBusy();
+        }
+    }
+
+    @Override
+    public CompletableFuture<LeaseMeta> getPrimaryReplica(ReplicationGroupId replicationGroupId, HybridTimestamp timestamp) {
+        if (!busyLock.enterBusy()) {
+            return failedFuture(new NodeStoppingException("Component is stopping."));
+        }
+        try {
+            // There's no sense in awaiting previously detected primary replica more than lease interval.
+            return awaitPrimaryReplica(replicationGroupId, timestamp).orTimeout(longLeaseInterval, TimeUnit.MILLISECONDS);
```

Comment:
Because it's maximum reasonable time to await for lease appearance based on lease propagation.
First of all, from the one hand leaseInterval should be gt the MAX_CLOCK_SKEW by design thus we've safe with clock drift here, on the other hand if distribution duration will be more than leaseInterval we may see an inappropriate gap.
Another approach with awaiting ms safe time, seems to be to invasive, code will become ugly, so I've finished with await bounding with given lease interval.

Reply:
I disagree that we can use long lease interval to limited waiting time. Timeout transaction is possible to use here, if I correctly understand how it is used in production, but the limitation is imposed externally.
Here, safe time is only possible to use. If it is not possible for now, you will be able to write TODO.

Reply:
Consider the following example:
- longLeaseInterval is 2 min
- cluster is experiencing some problems
- user runs very important transaction with a timeout 10 min, they are aware that there are some problems but have nothing to do with it right now, that's why the transaction timeout is rather long
- transaction starts, successfully performs some operation within, let say, 1 minute, and is going to be committed with commit timestamp 1000
- commit operation hangs for 2 minutes and then the transaction is rolled back at timestamp 1120 (let max clock skew be 0), because it was unable to get lease for timestamp 1000 withing this time interval. Actually lease for the interval (500, 1002) was received on corresponding node at local time 1121, but the transaction is already rolled back at this time.

So we have the transaction with timeout 10 minutes that was committed by user but rolled back by cluster after ~3 minutes. It is actually implicit commit timeout, which serves also as implicit and non-configurable transaction timeout. This looks at least weird from the user's point of view.

I understand that longLeaseInterval is long enough to count it as a reasonable time of data propagation, but the consequence is that we can roll back any transaction if some of its operations exceeds "reasonable" time which is not related to transaction timeout, and therefore, arbitrarily roll back any transaction before it reaches its timeout. This behavior impacts user experience, seems to be invalid and shouldn't exist.

So, the transaction timeout instead of safe time seems to be more correct here.

Reply:
To use safe time is absolutely correct, but any timeout cannot be used as a replacement, because any timeout might be violated. Time to spread information across cluster does not have time limitation.
We have only guaranties are related to clock of cluster, but this cannot be applied to safe time. Safe time on different nodes might be arbitrary different. Hence, we cannot to replace safe time check by checking timeout.

Reply:
> Timeout transaction is possible to use here
Nope

> To use safe time is absolutely correct
Yes. I've reworked getPrimaryReplica to use the safe time. And created a ticket in order to fix the race in meta storage safe time publication https://issues.apache.org/jira/browse/IGNITE-19532 @SammyVimes is notified and hopefully will fix it within next week. It worth to mention that I've run corresponding test 1000 times - no failures, so I event didn't mute it with the issue.